tests/unit/s2n_client_hello_test.c

```diff
@@ -1953,6 +1961,138 @@ int main(int argc, char **argv)
         };
     };
 
+    /* Test: Client Hello is slightly less than 64KB */
+    {
+        DEFER_CLEANUP(struct s2n_config *client_config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_NOT_NULL(client_config);
+
+        DEFER_CLEANUP(struct s2n_config *server_config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_NOT_NULL(server_config);
+
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, chain_and_key));
+
+        struct s2n_cipher_suite *test_cipher_suites[REDUCED_CIPHER_SUITE_COUNT] = { 0 };
+
+        for (int i = 0; i < REDUCED_CIPHER_SUITE_COUNT; i++) {
+            test_cipher_suites[i] = &s2n_rsa_with_aes_128_gcm_sha256;
+        }
+
+        const struct s2n_cipher_preferences test_cipher_suites_preferences = {
+            .count = s2n_array_len(test_cipher_suites),
+            .suites = test_cipher_suites,
+        };
+
+        struct s2n_security_policy test_security_policy = {
+            .minimum_protocol_version = S2N_TLS12,
+            .cipher_preferences = &test_cipher_suites_preferences,
+            .kem_preferences = &kem_preferences_null,
+            .signature_preferences = &s2n_signature_preferences_20240501,
+            .ecc_preferences = &s2n_ecc_preferences_20240501,
+            .rules = {
+                    [S2N_PERFECT_FORWARD_SECRECY] = true,
+            },
+        };
+
+        EXPECT_SUCCESS(s2n_config_disable_x509_verification(client_config));
+
+        DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client);
+        EXPECT_SUCCESS(s2n_connection_set_config(client, client_config));
+        EXPECT_SUCCESS(s2n_connection_set_blinding(client, S2N_SELF_SERVICE_BLINDING));
+        client->security_policy_override = &test_security_policy;
+
+        DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server);
+        EXPECT_SUCCESS(s2n_connection_set_config(server, server_config));
+        EXPECT_SUCCESS(s2n_connection_set_blinding(server, S2N_SELF_SERVICE_BLINDING));
+        server->security_policy_override = &test_security_policy;
+
+        DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+        EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+        EXPECT_OK(s2n_connections_set_io_stuffer_pair(client, server, &io_pair));
+
+        EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server, client, SERVER_HELLO));
+
+        /* handshake.io shouldn't be tainted after sending and receiving large client hello */
+        EXPECT_TRUE(client->handshake.io.tainted == 0);
+        EXPECT_TRUE(server->handshake.io.tainted == 0);
```

Comment:
This test isn't a defense against tainting the stuffer. You've got a very basic handshake here, and any of the optional extensions could still taint the stuffer. We got ?lucky? this time that the extension that tainted the stuffer was a common one. But as an example: how do you know that this handshake actually sent an ecc key share? You're negotiating TLS1.2-- it could be doing RSA kex (I don't think it is tho, RSA should not be the first option in default). 

If we actually want to defend against tainted stuffers, we're going to need to be more thorough here.